src/base58.rs

```diff
@@ -0,0 +1,441 @@
+// https://github.com/bitcoin/bitcoin/blob/master/src/base58.cpp
+// However, I tried not to look at the code so much, and instead tried to derive
+// some things myself.
+
+use crypto::{digest::Digest, sha2};
+
+const ALPHABET: &'static [u8; 58] = b"123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz";
+const CHAR_ALPHABET_INDEX_MAP: &'static [i8; 256] = &[
+    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+    -1, 0, 1, 2, 3, 4, 5, 6, 7, 8, -1, -1, -1, -1, -1, -1, -1, 9, 10, 11, 12, 13, 14, 15, 16, -1,
+    17, 18, 19, 20, 21, -1, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, -1, -1, -1, -1, -1, -1, 33,
+    34, 35, 36, 37, 38, 39, 40, 41, 42, 43, -1, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56,
+    57, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+];
+
+fn double_sha256(payload: &[u8]) -> [u8; 32] {
+    let mut final_hash: [u8; 32] = [0; 32];
+    let mut hasher = sha2::Sha256::new();
+    hasher.input(payload);
+    hasher.result(&mut final_hash);
+    hasher.reset();
+    hasher.input(&final_hash);
+    hasher.result(&mut final_hash);
+
+    final_hash
+}
+
+#[derive(Debug, PartialEq)]
+enum FromBase58Error {
+    InvalidCharacter,
+}
+
+trait ToBase58 {
+    fn to_base58(&self) -> String;
+}
+
+trait FromBase58 {
+    fn from_base58(&self) -> Result<Vec<u8>, FromBase58Error>;
+}
+
+impl ToBase58 for [u8] {
+    // After all the initial zeros are counted, what follows is just long
+    // division.
+    fn to_base58(&self) -> String {
+        let mut zero_counter: usize = 0;
+        let mut encoding_flag = false;
+
+        let mut b58_bytes: Vec<u32> = Vec::new();
+
+        let mut first_b58_byte_iter = true;
+        let mut size = 0;
+        for &byte in self.iter() {
+            if !encoding_flag {
+                zero_counter += (byte == 0) as usize;
+                if byte == 0 {
+                    continue;
+                }
+                encoding_flag = true;
+            }
+
+            let mut carry = byte as u32;
+            let mut i = 0;
+
+            while carry != 0 || i < size {
+                if first_b58_byte_iter {
+                    b58_bytes.push(0);
+                    first_b58_byte_iter = false;
+                } else {
+                    if let Some(b58_byte) = b58_bytes.get(i) {
+                        carry += b58_byte << 8;
```

Comment:
Why use `<< 8` instead of `* 256`?

I think it's more clear when someone is reading the spec.

Both probably have the same result after compilation (I **think**).

Reply:
I used `<< 8` because I was trying to do everything using only bitwise operators just for curiosity haha I forgot to removed the `<< 8` when I decided to use arithmetic operators instead, since bitwise operators leave the code really confusing if performance is not an issue.

And yea, I also think the compiler optimizes `* 256` to `<< 8` :)

I will change to `* 256`.

Reply:
It indeed does :)
If you go to https://godbolt.org/ and paste
```rust
pub fn m(a: u32) -> u32 {
    let b = a * 256;
    b
}
```

and then add `-O` as a compiler flag to `rustc`, the compiled assembly code is
```asm
example::m:
        mov     eax, edi
        shl     eax, 8
        ret
```

where `shl` is the left shift instruction.